Add blog article: Howto migrate a real-life batch pipeline from the DataSet API to the DataStream API

echauchot opened the pull request:
Cross posting to official flink blog from https://echauchot.blogspot.com/2022/11/flink-howto-migrate-real-life-batch.html. This blog article was already submitted to flink community for validation before publishing to my personal blog.

R: @zentol

echauchot committed:
Add blog article: Howto migrate a real-life batch pipeline from the DataSet API to the DataStream API

echauchot committed:
Add twitter handle

zentol commented on content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html:
```diff
@@ -0,0 +1,1093 @@
+
+<!DOCTYPE html>
+<html lang="en" dir=>
```
Please remove the .html files from the PR. Let the committer generate these once the merge is happening.

echauchot replied:
Yes I know. For some reasons I messed up my last commit's `git add`. That being said I understand why `content` directory is not in the git ignore, it is because you commit it as part of the website rebuild before merging right ?

echauchot replied:
done

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+
+## Migrating the DataSet pipeline to a DataStream pipeline
+
+There is no point in going through all of the code which is
+available [here](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java)
+. We will rather focus on some key areas of the migration. The code is based on the latest release
```
```suggestion
Instead of going through all of the code which is
available [here](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java)
 we will rather focus on some key areas of the migration. The code is based on the latest release
```

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+a bounded source, at some point the watermark will advance to +INFINITY (Long.MAX_VALUE) and will
+thus cross the end of the time window and consequently fire the trigger and output the records.
+
+Now that we have a working triggering, we just need to call a standard join with the  _Row::join_
```
```suggestion
Now that we have a working triggering, we need to call a standard join with the  _Row::join_
```

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+the [keyBy()](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/datastream/DataStream.html#keyBy-org.apache.flink.api.java.functions.KeySelector-)
+method. An aggregation downstream will be applied on elements with the same key exactly as
+a [GroupReduceFunction](https://nightlies.apache.org/flink/flink-docs-release-1.12/api/java/org/apache/flink/api/common/functions/GroupReduceFunction.html)
+would have done on a DataSet unless it will not materialize the collection of data. The summing
```
> unless it will not materialize the collection of data

Not sure what this is referring to.

echauchot replied:
Wording issue. I meant "except  it will not materialize the collection of data".
Correct me if I'm wrong but DataSet#groupBy() shuffles the data per key and materializes the data belonging to the same key in memory (in a collection) whereas DataStream#keyBy() doesn't do this materialization.

zentol replied:
> DataSet#groupBy() [...] materializes the data belonging to the same key in memory

That depends on what you're doing afterwards.
A reduce doesn't materialize everything and works similarly to a streaming reduce.

For a join, one/both sides are materialized in memory and/or spilled to disk.
Group Reduce materializes all data.

echauchot replied:
Makes sense. So in my case, as the following operator is a reduce, there will be indeed no materialization. Thanks for the details.

echauchot replied:
the details were added to the article, did you see it ?

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+the [onTimer()](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/functions/KeyedProcessFunction.html#onTimer-long-org.apache.flink.streaming.api.functions.KeyedProcessFunction.OnTimerContext-org.apache.flink.util.Collector-)
+callback.
+
+Another thing: to be able to use Flink state, we need to key the datastream beforeward even if there
```
```suggestion
Another thing: to be able to use Flink state, we need to key the datastream beforehand, even if there
```

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+to count them and stop at the set limit. Here also we need to key by a static key to be able to use
+the state API.
+The code resides in
+the [LimitMapper](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L232-L253)
```
This seems weird. It's super sketchy for a function to just close() itself.

Why not map records to "1" and apply a reduce function that maxes out at 100?

echauchot replied:
I agree it is sketchy. I did that to be sure to keep the order built in the previous step. That being said, I could replace by a `ProcessFunction` that does the limit. Indeed, I tested : there is no rebalancing of data (just a hash) between the order by step and the limit step in Flink execution graph.  

Also please note that I could have done much simpler code by mixing DataStream code with SQL/Table API especially using `LIMIT 100` and `ORDER BY` but the point was to try to code the equivalent of DataSet features with pure DataStream API to prepare the migration path for users (discover limitations such as [this one](https://issues.apache.org/jira/browse/FLINK-22587)) and also to compare the performance of both APIs.

echauchot replied:
Limit code was remade and article updated. Did you see it ?

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+that uses
+the [Flink SQL/Table API](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/overview/)
+in
+the [Query3ViaFlinkSQLCSV class](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkSQLCSV.java)
```
Let's point to a specific commit instead of the master branch

echauchot replied:
yes, I did that for all the other links in the other articles, thanks for catching this missing one !

echauchot replied:
done

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+Like for the order by migration, here also we need to adapt to the streaming semantics in which we
+don't have the whole data at hand. Once again, we use the state API. We apply a mapper that stores
+the incoming rows into
+a [ValueState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ValueState.html)
```
Why aren't you linking to the Flink javadocs at https://nightlies.apache.org/flink/flink-docs-release-1.16/api/java/?

echauchot replied:
because at the time I wrote this article I did not know that javadocs were also in nightlies and I forgot to change the links in this old article. Thanks for catching it, I'll update all the links in this article.

echauchot replied:
done

zentol reviewed (commented):
hmmm.......my impression is that some of things described here feel...sketchy? I don't know the batch execution mode enough for whether this is just the way things have to be done.

@alpinegizmo What do you think?

echauchot committed:
Fix alias and reformat

echauchot committed:
wording

Co-authored-by: Chesnay Schepler <chesnay@apache.org>

echauchot committed:
wording

Co-authored-by: Chesnay Schepler <chesnay@apache.org>

echauchot committed:
wording

Co-authored-by: Chesnay Schepler <chesnay@apache.org>

echauchot committed:
Fix javadoc urls

echauchot committed:
Reference commits instead of master branch in urls

echauchot committed:
update flink docs to 1.16 and not master

echauchot commented:
@zentol I have fixed a bug I mentioned earlier : missing windowing (it was not obvious to me that DataStream API in batch mode required windowing even if there is no event timestamps in the data, but nevermind). I also addressed your comments. 
PTAL.

echauchot committed:
fix typo

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -2,29 +2,39 @@
-aliases:
-- /news/2023/04/12/2023-04-12-howto-migrate-to-datastream.html
+  aliases:
+- /2023/04/12/2023-04-12-howto-migrate-to-datastream.html
```
we shouldnt need an alias; this is only required for older blog posts.

echauchot replied:
fixed

zentol commented on docs/content/posts/2023-04-12-howto-migrate-to-datastream.md:
```diff
@@ -0,0 +1,190 @@
+## The initial DataSet pipeline
+
+The pipeline we are migrating
+is [this one](https://github.com/echauchot/tpcds-benchmark-flink/blob/f342c1983ec340e52608eb1835e85c82c8ece1d2/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDataset.java)
+, it is a batch pipeline that implements the above query using the DataSet API
```
```suggestion
is [this](https://github.com/echauchot/tpcds-benchmark-flink/blob/f342c1983ec340e52608eb1835e85c82c8ece1d2/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDataset.java)
batch pipeline that implements the above query using the DataSet API
```

echauchot committed:
Remove alias

echauchot committed:
Add precisions on the materialization with keyBy

echauchot committed:
wording

Co-authored-by: Chesnay Schepler <chesnay@apache.org>

echauchot committed:
Fix formatting

echauchot committed:
Update text and links after recent code change.

echauchot commented:
@zentol finished updating PTAL

echauchot commented:
@zentol how does this article look like after last modifications ?

echauchot committed:
Update date of the article
